### PR TITLE
Upgrade to Ubuntu 16.04.2

### PR DESCRIPTION
**Pull request by jeffgeorge**
In #51 I proposed some pretty substantial upgrades:

- Upgrade to Ubuntu 16.04, to be in line with Acquia
- Switch from Apache `mod_php` to `php-fpm`, for better perf and because it's what Acquia does
  - Implement _two_ `php-fpm` pools, one for `php-5.6` and one for `php-7.0`
  - Enable users to control which pool is used via `config.rb`
- Expand our in-box tools, so Drupal sites can be built from the Guest rather than Host OS

. . so here's the PR that does all that.

## Upgrade Process

- Check out the `ubuntu-xenial` branch
- `vagrant destroy -f && vagrant up`
- Wait 5 minutes
- Check out your stuff

## Switching to PHP 7

- Crack open `config.rb`
- In a vhost you want to switch to PHP 7, add the following: `"php_version" => "7.0",`. Ex:

```ruby
drupal_sites = {
  "test" => {
    "host" => "test.vm",
    "path" => "test/docroot",
    "php_version" => "7.0",
  }
}
```

- `vagrant reload --provision`
- Check it out

## Comprehensive In-Box Build Tools

The box now has Composer, Drush, Bundler, the full Compass suite, NodeJS, NPM, and Yarn ready to use, so _certain proprietary Drush Make wrappers_ should work natively inside the box, should you not have a sufficient build environment on your Host. I've also exposed a new Vagrant setting to `config.rb` to help work with private repos in the box.

- Add `forward_ssh_agent = true` to your `config.rb`, just anywhere at the top or bottom, outside of the `drupal_sites` object
- `vagrant reload`
- Cool, now any SSH Agent stuff on your host will magically be available inside the VM
- `vagrant ssh`
- Once inside the box, `git clone` whatever private repos you need, install whatever build tool wrappers you need, and build into `/srv/www`
- These will end up on your Host OS in the /sites/ directory
- Once built, add your new site configurations to `config.rb` and `vagrant reload --provision`

## Weird Stuff to Be On The Lookout For

- Literally everything about PHP was ripped out and re-written. Twice. Please check to see that it's being configured with the right settings, that all the modules are still available etc.
- `php-fpm` may need some tweaking. I already had to bump the FastCGI Timeout to make sites load right the first time.
- idk what else ¯\\\_(ツ)\_/¯

### CODE REVIEW

**Comment by jeffgeorge** [comment]
Some miscellaneous benchmarking:

```
Environment:
- Vagrant 1.9.7
- Virtualbox 5.1.22
- macOS 10.12.16b
- Late 2013 MBP w/ 16G RAM

---

Precip 1.x, PHP 5.6, standard Drupal 7 site, caches warmed:

$ ab -n1000 -c10 https://ymt.vm/user/login

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:       13   38   7.0     38      70
Processing:    15   49   8.6     49      98
Waiting:       12   41   8.4     43      73
Total:         40   87  10.5     87     141

$ ab -n1000 -c10 https://ymt.vm/

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:       14   51  17.1     48     109
Processing:    16  102  49.2     91     258
Waiting:       11   78  40.4     71     227
Total:         46  153  58.0    147     332

---

Precip 2.x, PHP 5.6, standard Drupal 7 site, caches warmed:

$ ab -n1000 -c10 https://ymt.vm/user/login

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:       12   39   7.1     38      73
Processing:    21   53  10.1     51     118
Waiting:       14   46   8.9     44     108
Total:         45   92  13.1     89     165

$ ab -n1000 -c10 https://ymt.vm/

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        9   37   7.0     37      72
Processing:    22   55  10.0     55     109
Waiting:       12   46   9.7     45     104
Total:         44   92  12.5     91     180

---

Precip 2.x, PHP 7, standard Drupal 7 site, caches warmed:

$ ab -n1000 -c10 https://ymt.vm/user/login

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:       12   38   7.8     37      72
Processing:    12   48   9.6     48      88
Waiting:        9   40   9.5     41      72
Total:         43   85  11.4     86     133

$ ab -n1000 -c10 https://ymt.vm/

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:       14   38   9.1     38      74
Processing:    13   53  13.1     54      96
Waiting:        8   40  13.8     41      85
Total:         42   91  16.0     91     147

---

Provision time (excluding base box downloading):

Precip 1.x: 324.09 seconds
Precip 2.x: 384.91 seconds
```

TL,DR:
- Simple pages perform about the same.
- Complex pages will perform the same in best-case scenarios.
- Complex pages in _worst_-case scenarios may now perform nearly twice as good with PHP 5.6 (332ms vs 180ms) and well over 2.25x perf with PHP 7.0 (332ms vs 147ms).

**Comment by jeffgeorge** [comment]
@en-vcracan -- Could you please take a look at this branch? It should solve all the issues you were having with building Domino sites inside Precip.

**Comment by 4hmedSamir** [comment]
Works fine for me 👍

**Comment by amanaplan** [comment]
Precip 1.x vs. 2.x: Current Significant / Possibly Significant Differences in PHP 5.6 Build/Config

- Downgraded `zip` extension (1.x: 1.15.1; 2.x: 1.12.5)
  **Although 1.15.1 includes bug fixes and unbundles libzip, the [release notes](https://pecl.php.net/package-changelog.php?package=zip) don't seem to indicate any compatibility issues**
- Major `imagemagick` upgrade (1.x: 6.7.7-10 2017-03-14 Q16; 2.x: ImageMagick 6.8.9-9 Q16 x86_64 2017-05-26_
  **No significant API changes, but we should watch for potential issues**
- `_SERVER["HTTP_CACHE_CONTROL"]` default removed in 2.x
  **We may want to examine adding  a more sensible lower`max-age` setting for files not served by PHP, but not a showstopper**

File: README.md [hunk 1]
```diff
@@ -5,10 +5,10 @@
 
 ## What's Included?
 A full LAMP stack, and a few nice extras.
-- Ubuntu Server 14.04 LTS
+- Ubuntu Server 16.04.2 LTS
```
Comment (amanaplan) [inline README.md:8]: 👍🏻

**Review by amanaplan — APPROVED** [review]
🎉

**Comment by vitalie-cracan** [comment]
@jeffgeorge Thank you for this new version. I have tried the domino-build on it and it's working out of the box! There's was one kind of errorish message during the trial build of the moorings that I did:

```sh
> cd /srv/www
> domino-build --repo=git@github.com:Travelopia/domino-moorings.git --branch=develop --profile=moorings --dir=moorings --env=local

...
Drush make complete.
Moving downloaded profile files into the live profile.
rsync: rename "/srv/www/moorings/.git/objects/pack/.pack-5476c46db22f98d9940c45b4ec2c930d91ee0981.idx.i4uccs" -> ".git/objects/pack/pack-5476c46db22f98d9940c45b4ec2c930d91ee0981.idx": Operation not permitted (1)
rsync: rename "/srv/www/moorings/.git/objects/pack/.pack-5476c46db22f98d9940c45b4ec2c930d91ee0981.pack.YvQ0XI" -> ".git/objects/pack/pack-5476c46db22f98d9940c45b4ec2c930d91ee0981.pack": Operation not permitted (1)
rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1183) [sender=3.1.1]
  Copying the live profile to the new docroot.
  There were 2 permission failures while cleaning up.
    To correct manually:
      sudo rm -rf /srv/www/moorings
  Completing as an initial make...
  Copying the docroot into place.
  Restoring your original working directory files.
  No errors found, cleaning up temp data.
  Completed in 328.84s
```
The suggestion "To correct manually:" the very folder I was trying to build does not seem right though.

All in all this is a box we can certainly work with. I did have to do some tweaks. I will share them here, maybe someday someone will find it helpful:

* First time provisioning does not work. There are same errors, as before, related to downloading some `apt` packages. To fix it, I need to add at least this line somewhere, I choose shell/librarian.sh
`sudo apt-get update`

There are errors also related to keys for puppetlabs and mariadb repos. So, I actually added these:
```sh
# Puppetlabs
sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 7F438280EF8D349F
# Mariadb
sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys F1656F24C74CD1D8
sudo apt-get update
```

* Symbolic links on shared folders between Windows and Vagrant are a problem. To make it better, I add this to Vagrant file:
`vb.customize ["setextradata", :id, "VBoxInternal2/SharedFoldersEnableSymlinksCreate/v-root", "1"]`

That line cannot go just anywhere though, it needs to be in the `config.vm.provider :virtualbox do |vb|` block.

* I find that I need to run my Windows command prompt as Administrator, before vagrant up. It makes life easier.

* Since I spend a lot of time in Precip, I find that I want a nicer prompt and shell. So I install `zsh` with `oh-my-zsh`. 

Thanks a lot!

**Comment by jeffgeorge** [comment]
@en-vcracan -- I've pushed some new changes which I *believe* will cover everything. Please pull latest from the `ubuntu-xenial` branch and see how this all works for you:

> Symbolic links on shared folders between Windows and Vagrant are a problem.

[Got it](https://github.com/clwdev/precip/blob/d9c3cc09279774a095d16a091e02777389209a3f/Vagrantfile#L152-L155).

> First time provisioning does not work.

This sounds reminiscent of Issue #39, which we can no longer reproduce from Clearwater. As such, I've gone back and implemented our original plan for that: a partially pre-built version of Precip. You can toggle using the partially pre-built version with the `config.rb` setting:

```
use_packaged_precip = true
```

Just put it in `config.rb` on its own line, outside of the `drupal_sites` object.

When you `vagrant destroy -f && vagrant up` it'll swap to using [`clwdev/precip` instead of `clwdev/precip-16.04-base`](https://github.com/clwdev/precip/blob/d9c3cc09279774a095d16a091e02777389209a3f/Vagrantfile#L49-L58). `clwdev/precip` is effectively a box that has base packages installed and the various package signing keys baked in. This should increase the likelyhood that you'll be able to provision first-time.

> Since I spend a lot of time in Precip, I find that I want a nicer prompt and shell.

I've _also_ added a new "custom provisioner" to help you with that. If you create a new file in the `/shell` directory called `config.sh` it will automatically be pulled in and run as a post-provisioning step.

So, you can fill it with something like:

```
#!/bin/sh
export DEBIAN_FRONTEND=noninteractive
sudo apt-get -qq -y install zsh
sh -c "$(curl -fsSL https://raw.githubusercontent.com/robbyrussell/oh-my-zsh/master/tools/install.sh)"
sudo chsh -s /bin/zsh vagrant
```

and it'll automatically install `zsh` and `oh-my-zsh` for you, and set the `vagrant` user's shell to `zsh`. 

_Note: it'll probably complain a lot but as best I can tell it works._

This is also a method I would recommend for installing various build tools automatically, which I'll document over in the Domino Wiki shortly.

> I find that I need to run my Windows command prompt as Administrator, before vagrant up

This looks like a host issue, so there's not much I can do about that. Vagrant may just prefer running as an admin on Windows.

Finally:

> There's was one kind of errorish message during the trial build 

This looks more like an issue with the build script than Precip itself. I'll try to reproduce the issue and if I can I'll move it to an issue in that project. If it seems to be _working_ despite the big angry warning I'd just ignore it for the moment.

**Comment by vitalie-cracan** [comment]
@jeffgeorge many thanks, for all the updates. I will give it a try.

Meanwhile, is it ok to keep flagging issues on this thread? 

I have a minor one. Every time I run drush (or php) I get this warning:
`PHP Deprecated:  Comments starting with '#' are deprecated in /etc/php/5.6/cli/conf.d/99-overrides.ini on line 1 in Unknown on line 0`

I think `#` needs to be replaced with `;` in `puppet/precip/templates/php_overrides.erb`.

**Comment by vitalie-cracan** [comment]
@jeffgeorge 

Just tried the pre-built version. `vagrant up` times out, after a number of attempts to connect via ssh. I've tried multiple times, same error. This is not critical, we can work with the standard build, after a few hickups in initial provisioning.


```sh
DEBUG virtualbox_5_1:   - [1, "ssh", 2222, 22, "127.0.0.1"]
 INFO ssh: Attempting SSH connection...
 INFO ssh: Attempting to connect to SSH...
 INFO ssh:   - Host: 127.0.0.1
 INFO ssh:   - Port: 2222
 INFO ssh:   - Username: vagrant
 INFO ssh:   - Password? false
 INFO ssh:   - Key Path: ["C:/Users/vcracan/.vagrant.d/boxes/clwdev-VAGRANTSLASH-precip/2.0.0/virtualbox/vagrant_private_key"]
DEBUG ssh:   - connect_opts: {:auth_methods=>["none", "hostbased", "publickey"], :config=>false, :forward_agent=>false, :send_env=>false, :ks_only=>true, :paranoid=>false, :password=>nil, :port=>2222, :timeout=>15, :user_known_hosts_file=>[], :verbose=>:debug, :logger=>#<Logger:0c7d890 @progname=nil, @level=0, @default_formatter=#<Logger::Formatter:0x5c7d878 @datetime_format=nil>, @formatter=nil, @logdev=#<Logger::Loevice:0x5c7d848 @shift_size=nil, @shift_age=nil, @filename=nil, @dev=#<StringIO:0x5c7d8c0>, @mutex=#<Logger::LogDevice::LogDeviceMutex:0x5c730 @mon_owner=nil, @mon_count=0, @mon_mutex=#<Mutex:0x5c7d7e8>>>>, :keys=>["C:/Users/vcracan/.vagrant.d/boxes/clwdev-VAGRANTSLASH-precip/2.0/virtualbox/vagrant_private_key"]}
DEBUG ssh: == Net-SSH connection debug-level log START ==
DEBUG ssh: D, [2017-07-31T10:30:47.252693 #7064] DEBUG -- net.ssh.transport.session[2e3e420]: establishing connection to 127.0.0.1:2222
D, [2017-07-31T10:30:47.253670 #7064] DEBUG -- net.ssh.transport.session[2e3e420]: connection established
I, [2017-07-31T10:30:47.253670 #7064]  INFO -- net.ssh.transport.server_version[2e37e2c]: negotiating protocol version
D, [2017-07-31T10:30:47.253670 #7064] DEBUG -- net.ssh.transport.server_version[2e37e2c]: local is `SSH-2.0-Ruby/Net::SSH_4.1.0 i386-mingw32

DEBUG ssh: == Net-SSH connection debug-level log END ==
 INFO ssh: SSH not ready: #<Vagrant::Errors::SSHConnectionAborted: SSH connection was aborted! This usually happens when the machine is taki
too long to reboot or the SSH daemon is not properly configured on the VM.
First, try reloading your machine with `vagrant reload`, since a simple
restart sometimes fixes things. If that doesn't work, destroy your machine
and recreate it with a `vagrant destroy` followed by a `vagrant up`. If that
doesn't work, contact support.>
```

**Comment by jeffgeorge** [comment]
I fixed the simple issue with the php.ini comment, but still can't reproduce any of the issues you're experiencing during provisioning or booting the pre-packaged box. I'm starting to think it's specific to Windows.

**Comment by vitalie-cracan** [comment]
yes, I see. We'll have to leave it at this for now, since we need to turn our attention to Leboat and we do have a working Precip thanks to your help.

**Pull request merged**